[1.15] Some work on BEPs and Commands

kindlich opened the pull request:
See the commit messages of the two commits for more info.

kindlich committed:
Ignore casing for the CrT BEP prefixes

Makes `<recipetype:crafting>` and `<recipeType:crafting>` equal.
Fixes some issues where the camelCasing leads to errors that may not be
easy to spot

kindlich committed:
Merged dumpBracket and dump x Sources, Added CommandCollectionEvent

The command Collection Event is called after the CT commands were initialized but before they are added to the dispatcher.
This allows you to add subcommands properly.
Only dump commands added by the `@BracketDumper` (see below) won't be
present already because we add them later to spot conflicting
dump/subcommand names.

Now the `/ct dump xxx` command will also use `@BracketDumper` methods as sources
That was we don't need two handlers that do the same, one for
dumpBrackets and one for the dump command.
For this, the internal structure of how BracketDumpers were created was changed.
The annotation was changed to allow for changes in the generated
filename and Display Name (for the help command)

The rest of the commit is applying the changes to the existing command
dumpers

jaredlll08 commented on Crafttweaker_Annotation_Processors/src/main/java/com/blamejared/crafttweaker_annotation_processors/processors/validation/PreprocessorAnnotationValidationProcessor.java:
```diff
@@ -32,6 +31,27 @@ public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment
+                }
+                
+                if(!noArgPublicConstructorFound && anyConstructorFound){
+                    this.processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR, "Element is annotated as Preprocessor but has not public no-arg constructor!", element);
```
"but has not public"

should be
"but has no public"

kindlich replied:
Fixed ^^
Though the error will only come up at compiletime anyways

jaredlll08 commented on Crafttweaker_Annotation_Processors/src/main/java/com/blamejared/crafttweaker_annotation_processors/processors/validation/PreprocessorAnnotationValidationProcessor.java:
```diff
@@ -32,6 +31,27 @@ public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment
+                    if(enclosedElement.getKind() == ElementKind.CONSTRUCTOR) {
+                        anyConstructorFound = true;
+                        ExecutableElement constructor = (ExecutableElement) enclosedElement;
+                        final boolean noArg = constructor.getParameters().size() == 0;
```
Can't you call
```
isEmpty()
```

instead of ```
size() == 0
```
?

kindlich replied:
Lol, worked too long in csharp ^^

kindlich committed:
Fixed Typo and use isEmpty instead of length check

Too much C# I guess lol

kindlich committed:
Updated ZenCode

Added checks for special characters
Closes #999

kindlich committed:
Have the annotation processor project include the required projects

You'll still need to add reflections as well, I didn't want the maven file to be 10MB

kindlich committed:
Fix classpath reading for ZenWrapper
